### DISCUSSION TITLE
useAsyncData and $fetch different responses

### DISCUSSION THREAD

**Discussion by bairrada97** [nuxt/nuxt#17471] [Questions] [answered]
according to docs:

`const {data} = await useAsyncData('key', () => "APIendpoiint")`;

this will make data available to the `<template>` so we can render the response.
If I have a button where I call the same endpoint but this time use` $fetch()` the response doesnt come in raw.
this makes me not be able to reuse a function or variables.

Here's an example of what I try so far: 

```
const test = ref(null);

// call api function to be reusable every time I want to call this endpoint
const callAPIEndpoint = () => {
     return $fetch("ExternalApiEndpoint");
}

// make the api request on server
// on <template> I have to use test.data to be able to render the response cause useAsyncData returns a raw response
test.value = await useAsyncData('test1', () => {
    return callAPIEndpoint();
}

//make the api Request every time I click on a button
//on <template> I have to use test to be able to render the response
const clickButton = async () => {
  test.value = await callAPIEndpoint();
}

```

**Comment by wobsoriano** [comment]
`useAsyncData` returns `data`, `error`, `pending`, and `refresh`. Same as `useFetch`. So you really have to use `test.data` right?

And I think setting the result to another `ref` isn't a good practice? You can just use `refresh`.

```ts
const { data, refresh } = await useAsyncData('test1', () => {
    return callAPIEndpoint();
}
```

**Reply by bairrada97** [reply]
but isnt useAsyncData and useFetch only for page load? I have to use $fetch If I want to make api request on click of a button.
This will lead to if I want to render same object structure but different responses I cant use refresh or data.

If I have different endpoints but same object strucutre refresh will not work right? 
for example: 
I call endpoint1 using useAsyncData or useFetch
I call endpoint2 using $fetch

I attacht my test ref to both responses cause my <template> will render same object strucutre

**Reply by wobsoriano** [reply]
Check the options https://v3.nuxtjs.org/docs/usage/data-fetching/

**Comment by danielroe** [comment, marked as answer]
`useAsyncData` is a composable that you call once in your setup function. It does expose a `refresh` function you can call to refresh the data. But you wouldn't call it in a method or event handler. Whereas `$fetch`  can be called anywhere and at any time.

**Reply by bairrada97** [reply]
exactly, so I want call my useAsyncData on page load and then call $fetch anytime I want, reason is if I have different endpoints but with same object structure that will be rendered on DOM I cant use refresh and data from the useAsyncData, thats why I have to attach the response to a ref.

for example:
I call endpoint1 using useAsyncData or useFetch // response: {data: {title:"blabla"}}
I call endpoint2 using $fetch response: {title:"testStuff"}

On MY template I only want to use
```
<template>
  {{test}}
</template>
```

but Im forced to 
```
<template>
  {{test.data}} //on page load using useAsyncData
  {{test}} //using $fetch 
</template>
```

**Reply by wobsoriano** [reply]
> {{test.data}} //on page load using useAsyncData
> {{test}} //using $fetch

Like I said, `useAsyncData` returns an object and your final data is inside the `data` key.

**Reply by bairrada97** [reply]
exactly but $fetch doesnt return data key, my final object its already the response, only $fetch.raw

**Reply by wobsoriano** [reply]
What's stopping you from doing this?

Using plain `$fetch` without `useAsyncData` works like a normal `fetch API` so it doesn't have the `data` and `refresh` properties. Check this for reference https://github.com/unjs/ohmyfetch

```html
<script setup lang="ts">
  const yourFunction = () => {
    // If you call this outside `useAsyncData`, of course you will not have the data and refresh properties.
    return $fetch('http://api.open-notify.org/iss-now.json');
  }

  // Your data will be available in the data property if you use this.
  const { data, refresh } = await useAsyncData('users', () => yourFunction())
</script>

<template>
  <div>{{ data }}</div>
  <button @click="refresh">Refetch</button>
</template>
```

**Reply by bairrada97** [reply]
does this work with different endpoints?
reason Im not doing this its cause the endpoint inside "yourFunction" its different on page load and on click event, and your example use the same endpoint.

And because my endpoints are different that's why my initial idea was not use refresh property but $fetch

**Reply by wobsoriano** [reply]
> does this work with different endpoints?

Then use multiple `useAsyncData`?

```ts
const yourFunction = (url: string) => {
  return $fetch(url);
}

const {
  data: users,
  refresh: refreshUsers
} = await useAsyncData('users', () => yourFunction('/api/users'))

const {
  data: products,
  refresh: refreshProducts
} = await useAsyncData('products', () => yourFunction('/api/products'))
```

Better if you create a repro if this still doesn't answer your case :) https://codesandbox.io/

If you want to use a single `useAsyncData` with dynamic endpoint then just do something like this

```html
<script setup lang="ts">
  const id = ref(1)

  const {
    data,
    refresh
  } = await useAsyncData('characters', () => {
    return $fetch('https://rickandmortyapi.com/api/character/' + id.value)
  })

  const getNextChar = () => {
    id.value += 1
    refresh()
  }
</script>

<template>
  <div>{{ data }}</div>
  <button @click="getNextChar">Get next character</button>
  <button @click="refresh">Refetch</button>
</template>
```

**Reply by bairrada97** [reply]
```

<template>
  <div>{{ data }}</div>
  <button @click="clickButton">Refetch</button>
</template>

<script setup>
const endpoint = ref('https://random-data-api.com/api/stripe/random_stripe');
const yourFunction = (url) => {
  return $fetch(url);
};


const { data, refresh } = await useAsyncData('users', () =>
  yourFunction(endpoint.value)
);

const clickButton = () => {
  endpoint.value = 'https://random-data-api.com/api/address/random_address';
  refresh();
};
</script>

```

updated your solution for what I want and this works fine using the refresh

I prefer to use this approach instead 2 useAsyncData cause my second endpoint I only want to run on click and not on page load

**Reply by wobsoriano** [reply]
Great!

**Reply by bairrada97** [reply]
thanks for help, appreciated :)

**Reply by amirtbi** [reply]
Is useAsyncData a SSR and can be used for rendering data from getters in store?

**Reply by danielroe** [reply]
It's not clear what you're asking @amirtbi.
